bug 547630 (www-client/chromium-44.0.2376.0 Fails to Compile with +widevine)

gmt opened the pull request:
The "deepest" commit (0f15cb0 - github displays PR patches in reverse order relative to git, making a muddle out of PR patch-ordering discussion) includes the state-of-the-art collaboratively cooked-up recipe for chromium widevine enablement on gentoo.

The middle one (4b2faff) tries to implement the "TODO" from the ebuild and turn the REQUIRED_USE into a package.use.mask cascade.  I'm pretty unsure about this one, it needs some review.

The head commit (3c2a24c) updates the blurb of einfo that www-plugins/chrome-binary-plugins[widevine] spews to be a more clear, accurate and up-to-date.  The live gx86 behavior is quite confusing, encouraging the user to enable a local use-flag which no longer exists in the cited gx86 package.  0f15cb0 only partially fixes that; this goes the whole way).

floppym commented:
@gmt Thanks for the pull request.

I really do not want to be stuck maintaining a patch against the chromium source tree.

As I understood from the bug report, there is some solution which does NOT involve patching the chromium source code -- the only downside being that the version is not reported correctly in the plugins page.

Also, I do not think it is appropriate to have IUSE="+widevine". Please default it to off.

gmt commented:
On Mon, Aug 24, 2015 at 7:57 AM, Mike Gilbert <notifications@github.com>
wrote:

> I really do not want to be stuck maintaining a patch against the chromium
> source tree.
>
That's definitely the principal downside of Ari's approach.  I'll see if
there's any hope of getting his patch upstream.  From my vantage point, the
chromium-dev decision-making process is quite opaque, so, as far as I know,
they might be interested and able to take Ari's patch (which is clearly the
best solution to the problem currently available, maintainability issues
aside).

> As I understood from the bug report, there is some solution which does NOT
> involve patching the chromium source code -- the only downside being that
> the version is not reported correctly in the plugins page.
>
I suppose.  But I gotta admit it really squicks me.   As a compromise,
assuming chromium doesn't jump all over Ari's patch, forthwith, would you
be open to something like this one-liner, which simply declines to report
any version at all?:


http://bazaar.launchpad.net/~saiarcot895/chromium-browser/chromium-browser.wily.beta/view/head:/debian/patches/fix_building_widevinecdm_with_chromium.patch

> Also, I do not think it is appropriate to have IUSE="+widevine". Please
> default it to off.
>
Yes, sensible idea.  Probably a vestige of the older drop-kicked
implementation, or, if not, a bad call on my part.

I will rev; first I'll give Google a heads-up and a few days to merge Ari's
patch (unless they already had it), and await your thoughts about
fix_building_widevinecdm_with_chromium.patch.

At least that gets us down to just three contenders (all other strategies
discussed in the bug being pareto-dominated by them): Ari's touchdown,
Ubuntu's punt, or upstream's fumble.

gmt commented:
presumably above QA check was temporary result of putting my repo into an inconsistent state.  Apparently, if you rebase your repo, github can automagically update your pull request (but... how on earth can it know which commits in my modified repo history correspond to the original PR commits!? :bowtie: <- :boom:... head assplode!  Anyhow whatever mechanism it used it guessed right).  I think it should be rectified now, will the CI bot let us know automatically?

gmt commented:
BTW, the automagically PR update addresses Mike's concerns:

ike Gilbert <notifications@github.com> wrote:

> I really do not want to be stuck maintaining a patch against the chromium
> source tree.
>
I replaced it with a trivial patch.

> As I understood from the bug report, there is some solution which does NOT
> involve patching the chromium source code -- the only downside being that
> the version is not reported correctly in the plugins page.
>
No, some patching is necessary.  I think this represents the minimum to
make it compile.

> Also, I do not think it is appropriate to have IUSE="+widevine". Please
> default it
>
fixed.

NB: I have only tested building with +widevine so far in this new rev.
I'll test building without and if there is any problem with it, report back
(otherwise, I presume it's fine).

atrigent commented:
> No, some patching is necessary.  I think this represents the minimum to make it compile.

The [latest thing][1] I posted in the bug does allow you to go without any patch, by setting WIDEVINE_CDM_VERSION_STRING in CPPFLAGS. However, as I noted in that comment, this would not be my preferred method.

[1]: https://bugs.gentoo.org/show_bug.cgi?id=547630#c76

phajdan commented:
Can you change things in a way that the necessary patches can be landed upstream? A simpler patch is better than a bigger patch, but if it's a perpetual patch that's not optimal either.

Also, could you just change the latest ebuild (i.e. for the dev channel)? That'd help make the pull request smaller and less repetitive. It'd also better reflect the way we maintain the chromium ebuild, i.e. land any enhancement changes in dev channel and wait for that to be promoted to beta and then stable.

gmt commented:
On Fri, Sep 11, 2015 at 3:59 AM, phajdan <notifications@github.com> wrote:

> Can you change things in a way that the necessary patches can be landed
> upstream?


No.  See above.

phajdan commented:
Could you comment more why (or what specifically are you referring to - it's not obvious to me)? I see an alternative approach e.g. in https://bugs.gentoo.org/show_bug.cgi?id=547630#c76 .

gmt commented:
On Fri, Sep 11, 2015 at 3:59 AM, phajdan <notifications@github.com> wrote:

> Also, could you just change the latest ebuild (i.e. for the dev channel)?


Maybe.  I run ::beta.  It takes me a few hours to test any build on my
workstation.  I'll probably need to cobble together a distcc farm if I'm
going to keep on top of more than one branch.  Not saying that as some kind
of rhetorical flourish -- I'm actually planning to do just that if I can
find the time tomorrow, I need one anyhow lately.

I'm surprised to hear this request, though, as I figured maintainable
full-tree rejiggerment was half the win of structuring something like this
as a PR instead of as a bunch of bgo attachments...  well, w/e it's early
days, I'm sure all will become clear as we bumble forward and find the
pain-points.

gmt commented:
On Fri, Sep 11, 2015 at 4:17 AM, phajdan <notifications@github.com> wrote:

> Could you comment more why (or what specifically are you referring to -
> it's not obvious to me)? I see an alternative approach e.g. in
> https://bugs.gentoo.org/show_bug.cgi?id=547630#c76 .
>

Sorry to be curt but this issue has a
sprawling-ocean-of-almost-equivalent-proposals problem that causes S/N
problems, so hoping to avoid recreating the bugzilla mess here.

The original rev of this PR did exactly what you are hinting at, but Mike
doesn't like it above.  I've come to the conclusion that he is correct not
to.

bgo #547630 Comment 84 provides an institutional and a deontological
justification for why I rev'ed: the maintenance burden is a cost, but there
is afaics no known benefit to in-the-wild usership of the widevine cdm, so
the whole open question boils down to a bunch of fussy aesthetically
motivated feature-creep that no known end-users care about and which has no
hope of ever going upstream (see crbug.com/473866)

I started that process -- my bad -- hoping maybe I can end it, now that I
see the error of my former ways :)

floppym commented:
I'm working on a modified version of Greg's latest changes here, and plan to have them added to chromium-46 and 47 today.

floppym commented:
> Can you change things in a way that the necessary patches can be landed upstream?

To get this landed upstream, src/chrome/common/chrome_content_client.cc would need to be modified to not refer to the WIDEVINE_CDM_VERSION_STRING macro. That would probably mean moving the version information into libwidevinecdm.so itself and retrieving it at runtime.

Unfortunately a closed-source component, and so we can't provide a patch for that.

atrigent commented:
> To get this landed upstream, src/chrome/common/chrome_content_client.cc would need to be modified to not refer to the WIDEVINE_CDM_VERSION_STRING macro. That would probably mean moving the version information into libwidevinecdm.so itself and retrieving it at runtime.
>
> Unfortunately a closed-source component, and so we can't provide a patch for that.

Please see [the upstream bug][1] for my proposal for getting rid of WIDEVINE_CDM_VERSION_STRING. Changing the library itself is not necessary, but the official Chrome for Linux packages do need to be changed slightly, and I'm waiting on someone in the know to tell me how to do this.

[1]: https://code.google.com/p/chromium/issues/detail?id=473866#c5
